Libplanet/Blockchain/BlockChain.cs

```diff
@@ -539,23 +539,39 @@ internal HashDigest<SHA256> FindBranchPoint(BlockLocator locator)
             {
                 _rwlock.EnterReadLock();
 
-                // FIXME Depending on the amount of Index stored in the `Store`,
-                //       it can use the memory excessively.
-                //       we should find alternative approach.
-                ImmutableHashSet<HashDigest<SHA256>> indices = Store
-                    .IterateIndex(Id.ToString())
-                    .ToImmutableHashSet();
+                IEnumerable<HashDigest<SHA256>> indexes = Store.IterateIndex(
+                    Id.ToString()
+                );
+
+                var hashes = locator.ToImmutableArray();
+                var hashesIndex = hashes.Length - 1;
 
                 // Assume locator is sorted descending by height.
-                foreach (HashDigest<SHA256> hash in locator)
+                foreach (HashDigest<SHA256> index in indexes)
```

Comment:
It seems to assume that `indexes` are ordered in a particular order, right?

Reply:
Yes, I assumed that `IterateIndex` would return indexes in order

Reply:
if so, it would be better if `Store.IterateIndex()` returns `IOrderedEnumerable` and be documented too.

Reply:
`IStore.IterateIndex` can't return `IOrderedEnumerable` because `IOrderedEnumerable` is not an iterator interface type. In order to maintain the function of it, we should use `IEnumerable` and document it.

Reply:
> `IOrderedEnumerable` is not an iterator interface type.

AFAIK [`IOrderedEnumerable<T>`][1] extends [`IEnumerable<T>`][2].

[1]: https://docs.microsoft.com/en-us/dotnet/api/system.linq.iorderedenumerable-1
[2]: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1

Reply:
In [docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/yield#iterator-methods-and-get-accessors) from microsoft,
```
yield (C# Reference)
...

The return type must be IEnumerable, IEnumerable<T>, IEnumerator, or IEnumerator<T>.
```

Reply:
Indeed [it seems tricky.][1]  We'd better to mention this in the docs.

[1]: https://stackoverflow.com/questions/14404139/is-it-possible-to-turn-an-ienumerable-into-an-iorderedenumerable-without-using-o